### ISSUE TITLE
[css-animations-2][css-style-attr] Anonymous inline keyframes (e.g. in a style attribute)

### ISSUE THREAD

**Issue by noamr** [w3c/csswg-drafts#10951]
Currently to give an element a keyframe animation, the author has to declare the `@keyframes` rule in one place, give them a name, and then assign the keyframes to an `animation-name`.
This is unnecessarily tedious in some cases where the keyframes have a 1:1 relationship with an element or selector, and also restrict their use in inline styles.

Asking that we find a syntax to "shortcut" the keyframes directly into `animation-name`. Perhaps:

```css
animation-name: @keyframes { ... }
```

**Comment by birtles** [comment]
Just out of curiousity, is this specifically for SSR use cases?

(I'm asking since so far I've mostly seen that for one-off keyframes the Web Animations API is the easiest route, but I can see how in SSR situations generating CSS would be better.)

**Comment by ydaniv** [comment]
> Just out of curiousity, is this specifically for SSR use cases?
>
> (I'm asking since so far I've mostly seen that for one-off keyframes the Web Animations API is the easiest route, but I can see how in SSR situations generating CSS would be better.)

Yes, it you're not working with an effects library, either external or internal, then this could be a bit more verbose, but perhaps more simple to maintain.

**Comment by kizu** [comment]
I was thinking about the same recently: [mastodon post](https://front-end.social/@kizu/113211471973334178), and @bramus also commented there about it.

I wonder if we need to use the `@` there, or if it could be just a `keyframes()` function? With the implication that you would be able to assign it to a variable, and pass down the scope, exposing this specific animation for any children.

My use cases are mostly scroll-driven animations: they can often be useful for one-off properties, and often all you need a simple `keyframes { to { --foo: 1 } } `, but with the `--foo` that can be different for different cases and elements. Right now, we have to create many very similar custom `@keyframes`, but being able to use them inline would be great, as you won't have to think about naming, and they won't be reused anywhere else.

**Comment by noamr** [comment]
> I was thinking about the same recently: [mastodon post](https://front-end.social/@kizu/113211471973334178), and @bramus also commented there about it.
> 
> I wonder if we need to use the `@` there, or if it could be just a `keyframes()` function? With the implication that you would be able to assign it to a variable, and pass down the scope, exposing this specific animation for any children.

Yea it should probably be `keyframes()` rather than `@keyframes`, but in either case the value should inherit and behave like a regular value. I was going with `@` at first because the insides are pretty rich for a function.

**Comment by noamr** [comment]
Some idea, what if `keyframes()` was a per attribute function?

So instead of:
```css
@keyframes anim {
  from { opacity: 0.3; translate: 100px; }
  50% { opacity: 0.5; translate: 200px; }
  to { opacity: 1; translate: 1000px }
}

.me { animation-name: anim }
```

you can do something equivalent to `if`

```css
.me {
  animation-name: auto;
  opacity: keyframes(from: 0.3; 50%: 0.5; to:1; }
  translate: keyframes(from: 100px; 50%: 200px; to: 1000px; }
}
```

Perhaps with an "else" like semantic for the base value.
This would naturally work in inline styles.

**Comment by noamr** [comment]
cc @LeaVerou @fantasai @tabatkins @flackr WDYT?

**Comment by LeaVerou** [comment]
+1 on the pain point, don't love the syntax.
The actual pain point is that you need to create a global (tree-scoped) identifier every time you want an animation.
That doesn't need a parallel syntax, it just needs scoped names!
What if we could nest a `@keyframes` rule in another rule and use hygienically rewritten names? Then you need to put significantly less thought in picking a name that won't clash.
That would also allow mixins to provide animations, which is a common need.

While we're at it, a pain point that web component authors brought up to us recently is that WC consumers cannot pass animations to replace default ones due to the current scoping rules. I wonder if we could feed two birds with one scone here. 🤔

**Comment by noamr** [comment]
> +1 on the pain point, don't love the syntax. The actual pain point is that you need to create a global (tree-scoped) identifier every time you want an animation. That doesn't need a parallel syntax, it just needs scoped names! 

Why would I need a name at all? I just want to set keyframes on a particular element. Naming feels like tax...

**Comment by flackr** [comment]
> Some idea, what if `keyframes()` was a per attribute function?
> 
> So instead of:
> 
> @keyframes anim {
>   from { opacity: 0.3; translate: 100px; }
>   50% { opacity: 0.5; translate: 200px; }
>   to { opacity: 1; translate: 1000px }
> }
> 
> .me { animation-name: anim }
> you can do something equivalent to `if`
> 
> .me {
>   animation-name: auto;
>   opacity: keyframes(from: 0.3; 50%: 0.5; to:1; }
>   translate: keyframes(from: 100px; 50%: 200px; to: 1000px; }
> }
> Perhaps with an "else" like semantic for the base value. This would naturally work in inline styles.

You haven't specified the timing, easing, fill modes, repeat, timelines, ranges, etc of those animations. Would it all go in the keyframes function?

Using the animation property would also work in inline styles and would provide a more model where multiple properties are animating in the same animation. If you have a keyframes function per property then it would be hard to coordinate associated property changes.

My main concern is the complexity of animation parsing and matching. Parsing is pretty self evident, but there is also the issue that whenever the animation-name property (the property where these keyframes would slot in) changes, we evaluate whether the animation name changed from what it was before. If instead of a name we have a whole description of the animation that comparison is going to be more expensive. It's certainly doable, just something to keep in mind, e.g.

```js
elem.style.animation = `
    @keyframes { 0% { opacity: 0; transform: translateY(-2rem); } } 5s,
    @keyframes { 0% { color: red; } 100% { color: green; } } 2s;`;
// Some time passes
elem.style.animation = `
    @keyframes { 0% { opacity: 0; transform: translateY(-1.5rem); } } 5s,
    @keyframes { 0% { color: red; } 100% { color: green; } } 2s;`;
```

This needs to detect that the second animation is the same in both, but the first animation changed so it is treated as a brand new animation starting over from the beginning.

**Comment by noamr** [comment]
> > Some idea, what if `keyframes()` was a per attribute function?
> > So instead of:
> > [@Keyframes](https://github.com/Keyframes) anim {
> > from { opacity: 0.3; translate: 100px; }
> > 50% { opacity: 0.5; translate: 200px; }
> > to { opacity: 1; translate: 1000px }
> > }
> > .me { animation-name: anim }
> > you can do something equivalent to `if`
> > .me {
> > animation-name: auto;
> > opacity: keyframes(from: 0.3; 50%: 0.5; to:1; }
> > translate: keyframes(from: 100px; 50%: 200px; to: 1000px; }
> > }
> > Perhaps with an "else" like semantic for the base value. This would naturally work in inline styles.
> 
> You haven't specified the timing, easing, fill modes, repeat, timelines, ranges, etc of those animations. Would it all go in the keyframes function?

No, it will use the existing attributes. This only replaces `animation-name`.

> Using the animation property would also work in inline styles and would provide a more model where multiple properties are animating in the same animation. If you have a keyframes function per property then it would be hard to coordinate associated property changes.

Sure, but the existing model of going by keyframe already exists. I'll be happy if we found an inline solution to that, but I wanted to bring up an alternative.

> My main concern is the complexity of animation parsing and matching. Parsing is pretty self evident, but there is also the issue that whenever the animation-name property (the property where these keyframes would slot in) changes, we evaluate whether the animation name changed from what it was before. If instead of a name we have a whole description of the animation that comparison is going to be more expensive. It's certainly doable, just something to keep in mind, e.g.
> 
> elem.style.animation = `
>     @keyframes { 0% { opacity: 0; transform: translateY(-2rem); } } 5s,
>     @keyframes { 0% { color: red; } 100% { color: green; } } 2s;`;
> // Some time passes
> elem.style.animation = `
>     @keyframes { 0% { opacity: 0; transform: translateY(-1.5rem); } } 5s,
>     @keyframes { 0% { color: red; } 100% { color: green; } } 2s;`;
> This needs to detect that the second animation is the same in both, but the first animation changed so it is treated as a brand new animation starting over from the beginning.

Right, I concede there's some implementation complexity here but I wanted to check if people want this first.
Though I'm also open to having anonymous keyframes with some syntax without transposing the model.

**Comment by LeaVerou** [comment]
> > +1 on the pain point, don't love the syntax. The actual pain point is that you need to create a global (tree-scoped) identifier every time you want an animation. That doesn't need a parallel syntax, it just needs scoped names!
> 
> Why would I need a name at all? I just want to set keyframes on a particular element. Naming feels like tax...

It _is_ tax. The idea is that if the names are private, the tax is significantly reduced.

I'm all for making it possible to define structured values inline without indirection, but this should not be solved ad hoc for `@keyframes` alone — they are not the only case that needs it (though they may well be the most pressing).

**Comment by noamr** [comment]
> > Why would I need a name at all? I just want to set keyframes on a particular element. Naming feels like tax...
> 
> It _is_ tax. The idea is that if the names are private, the tax is significantly reduced.

I'm seriously interested in making that tax 0 if possible :)
 
> I'm all for making it possible to define structured values inline without indirection, but this should not be solved ad hoc for `@keyframes` alone — they are not the only case that needs it (though they may well be the most pressing).

More examples that have a clear use case?
We do something similar already for `view()` so I think we can be consistent with that as much as possible

**Comment by LeaVerou** [comment]
> > > Why would I need a name at all? I just want to set keyframes on a particular element. Naming feels like tax...
> > 
> > 
> > It _is_ tax. The idea is that if the names are private, the tax is significantly reduced.
> 
> I'm seriously interested in making that tax 0 if possible :)

0 is definitely better than ε, but it seems like diminishing returns. 
OTOH expanding hygienic rewrite to other identifiers could be quite powerful.

> > I'm all for making it possible to define structured values inline without indirection, but this should not be solved ad hoc for `@keyframes` alone — they are not the only case that needs it (though they may well be the most pressing).
> 
> More examples that have a clear use case? We do something similar already for `view()` so I think we can be consistent with that as much as possible

- `@position-try` 
- `@counter-style` 
- `@font-face` (for one-off aliases)
- `@font-feature-values` 
- `@font-palette-values`

If we stretch it, to include temporary reuse of logic without defining a whole tree-scoped name, it could also include:
- `@function` 
- `@mixin` for the same reason
- `@custom-media`
- `@custom-supports`
